src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+        })?;
+
+        Ok(Self {
+            trace_path,
+            wal_path,
+            batches: Vec::new(),
+            compaction: None,
+        })
+    }
+
+    /// Remove all on-disk data for this trace.
+    ///
+    /// Importantly, we also delete the WAL directory here (the WAL writer
+    /// only gets to add new files and can't do anything else)
+    fn destroy(self) -> Result<(), anyhow::Error> {
+        fs::remove_dir_all(&self.trace_path).with_context(|| {
+            format!(
+                "failed to remove trace directory {}",
+                self.trace_path.display()
+            )
+        })?;
+        fs::remove_dir_all(&self.wal_path).with_context(|| {
+            format!("failed to remove wal directory {}", self.wal_path.display())
+        })?;
+
+        Ok(())
+    }
+
+    /// Checks if there are finished WAL segments and if so, forms them into batches.
+    fn consume_wal(&mut self) -> Result<(), anyhow::Error> {
+        let finished_segments = self.find_finished_wal_segments()?;
+
+        for segment in finished_segments {
+            let batch = Batch::create(&segment, &self.trace_path)?;
+            self.batches.push(batch);
+            fs::remove_file(&segment).with_context(|| {
+                format!(
+                    "failed to remove consumed wal segment {}",
+                    segment.display()
+                )
+            })?;
+        }
+
+        Ok(())
+    }
+
+    fn find_finished_wal_segments(&self) -> Result<Vec<PathBuf>, anyhow::Error> {
+        lazy_static! {
+            static ref FINISHED_WAL_SEGMENT_REGEX: Regex =
+                Regex::new("^log-[0-9]+-final$").unwrap();
+        }
+
+        let mut segments = read_dir_regex(&self.wal_path, &FINISHED_WAL_SEGMENT_REGEX)?;
+        segments.sort_by_key(|segment| {
+            segment
+                .to_str()
+                .unwrap()
+                .split('-')
+                .nth(1)
+                .unwrap()
+                .parse::<usize>()
+                .unwrap()
+        });
+
+        Ok(segments)
+    }
+
+    fn find_unfinished_wal_segment(&self) -> Result<PathBuf, anyhow::Error> {
+        lazy_static! {
+            static ref UNFINISHED_WAL_SEGMENT_REGEX: Regex = Regex::new("^log-[0-9]+$").unwrap();
+        }
+
+        let mut segments = read_dir_regex(&self.wal_path, &UNFINISHED_WAL_SEGMENT_REGEX)?;
+        if segments.len() > 1 {
+            bail!(
+                "Expected only a single unfinished wal segment at {}. Found {}",
+                self.wal_path.display(),
+                segments.len()
+            );
+        }
+
+        if segments.len() == 0 {
+            bail!(
+                "Expected at least a single unfinished wal segment at {}. Found none.",
+                self.wal_path.display()
+            );
+        }
+
+        Ok(segments.pop().unwrap())
+    }
+
+    fn find_batches(&self) -> Result<Vec<Batch>, anyhow::Error> {
```

Comment:
The previous method names were clear enough that I understood them by name alone, but not this one!

Reply:
From a subsequent comment: this probably wants to think harder about the batches it can find, and discard any batches that are subsumed by other batches in the list. E.g. if you merged a batch and then failed before deleting their constituents.